### PR TITLE
feat: add type casting to `FlaxModelMixin`

### LINKED ISSUE #2068: Specifying dtype in from_pretrained does not convert the model's datatype
### Describe the bug

When I grab a model using "from_pretrained" and specifying the datatype, the returned model's parameters do not have that type.

This causes problems with flax text to image and flax dreambooth training scripts [here](https://github.com/huggingface/diffusers/blob/ce1c27adc85916a802db579df789d990fd14e8bc/examples/dreambooth/train_dreambooth_flax.py)
 and [here](https://github.com/huggingface/diffusers/blob/ce1c27adc85916a802db579df789d990fd14e8bc/examples/text_to_image/train_text_to_image_flax.py) since the "mixed_precision" argument does not convert the pretrained model, my TPU was running out of memory when training.

### Reproduction

`
from diffusers import FlaxStableDiffusionPipeline
import jax.numpy as jnp
unet, unet_params = FlaxUNet2DConditionModel.from_pretrained(
    "duongna/stable-diffusion-v1-4-flax", subfolder="unet", dtype=jnp.bfloat16
)
print(unet_params)`

returns
`
{conv_in': {'bias': Array([...] dtype=float32) } [...]}`

Expected result: The datatype for the models parameters are jnp.bfloat16

Actual result: The datatype for the models parameters are float32.

Possible solution: Can be fixed by using jax tree utils

`
def convert_dtype(xs, dtype): return jax.tree_util.tree_map(lambda x: jnp.asarray(x, dtype=dtype), xs)
`

### Logs

_No response_

### System Info

- `diffusers` version: 0.12.0.dev0
- Platform: Linux-5.13.0-1027-gcp-x86_64-with-glibc2.31
- Python version: 3.10.8
- PyTorch version (GPU?): 1.13.1+cu117 (False)
- Huggingface_hub version: 0.11.1
- Transformers version: 4.25.1
- Accelerate version: not installed
- xFormers version: not installed
- Using GPU in script?: No
- Using distributed or parallel set-up in script?: N/A

### PR DESCRIPTION
**Pull request by SauravMaheshkar**
# What does this PR do?

Fixes #2068 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Request for Review: @patrickvonplaten and @pcuenca

### CODE REVIEW

File: src/diffusers/models/modeling_flax_utils.py [hunk 1]
```diff
@@ -435,6 +435,7 @@ def from_pretrained(
             # make sure all arrays are stored as jnp.ndarray
             # NOTE: This is to prevent a bug this will be fixed in Flax >= v0.3.4:
             # https://github.com/google/flax/issues/1261
+        state = cls._cast_floating_to(state, dtype)
```
Comment (pcuenca) [inline src/diffusers/models/modeling_flax_utils.py:438]: @SauravMaheshkar did you forget to update `_cast_floating_to`? I don't think this works as it's not a class method in the current implementation.

File: src/diffusers/models/modeling_flax_utils.py [hunk 2]
```diff
@@ -435,6 +436,7 @@ def from_pretrained(
             # make sure all arrays are stored as jnp.ndarray
             # NOTE: This is to prevent a bug this will be fixed in Flax >= v0.3.4:
             # https://github.com/google/flax/issues/1261
```
Comment (pcuenca) [inline src/diffusers/models/modeling_flax_utils.py:438]: I'd remove the note about a flax bug if we end up doing the conversion on jnp arrays. And indent the comments correctly.

Comment (pcuenca) [inline src/diffusers/models/modeling_flax_utils.py:438]: I'd remove the note about a flax bug if we end up doing the conversion on jnp arrays. And indent the comments correctly, they got out of sync at some point.

File: tests/models/test_modeling_common_flax.py [hunk 3]
```diff
@@ -64,3 +67,104 @@ def test_deprecated_kwargs(self):
...
+        init_dict, inputs_dict = self.prepare_init_args_and_inputs_for_common()
+
+        # check if all params are still in float32 when dtype of computation is half-precision
+        model = self.model_class(**init_dict, dtype=jnp.float16)
+        params = model.init(inputs_dict["prng_key"], inputs_dict["sample"])
+        types = jax.tree_util.tree_map(lambda x: x.dtype, params)
+        types = flatten_dict(types)
+
+        for name, type_ in types.items():
+            self.assertEquals(type_, jnp.float32, msg=f"param {name} is not initialized in fp32.")
```
Comment (pcuenca) [inline tests/models/test_modeling_common_flax.py:81]: I'm confused, wasn't the purpose of the PR to change the dtype on load?

**Review by pcuenca — COMMENTED** [review]
Sorry, I'm afraid I'm a bit confused about this PR. I thought the purpose was to convert the `dtype` to whatever the user specified, but the tests seem to do something else: the user specifies the desired dtype but `test_default_params_dtype` verifies that it **didn't** change. Could you please help me understand how it's supposed to work?

- If we really want to change the `dtype`, then the `_cast_floating_to` invocation should be after the conversion to jnp arrays in the following line. Otherwise, this line will prevent the conversion from happening (which is the case now): https://github.com/huggingface/diffusers/blob/8e5921cac14c84a670b4e9d89d717b5356e5b9e9/src/diffusers/models/modeling_flax_utils.py#L72
- In addition, tests are failing for the VAE. Could you please verify them?

Thanks for iterating with us, I'd just like to understand this better :)

File: src/diffusers/models/modeling_flax_utils.py [hunk 4]
```diff
@@ -435,6 +436,7 @@ def from_pretrained(
...
+        state = cls._cast_floating_to(state, dtype)
```
[the remark this answers is not in the document]
Comment (SauravMaheshkar) [reply src/diffusers/models/modeling_flax_utils.py]: Yes, thank you. Fixed in 6803f8883568bd7e72bc0cf5c5cefb59755cbcf9

**Pull request closed without merging**